### ISSUE TITLE
Provide `deno check` as runtime API

### ISSUE THREAD

**Issue by jsejcksn** [denoland/deno#15478]
[`Deno.emit`](https://doc.deno.land/deno/unstable@v1.21.3/~/Deno.emit) has [come](https://github.com/denoland/deno/pull/8799) and [gone](https://github.com/denoland/deno/pull/14463), and plenty of [suggestions can be found pointing toward using `esbuild`](https://github.com/denoland/deno/issues?q=is%3Aissue+esbuild) to _transpile_ TypeScript source code at runtime, but there is now a lack of a runtime API for type-checking.

Previously, with `Deno.emit`, [type-related diagnostics](https://doc.deno.land/deno/unstable@v1.21.3/~/Deno.Diagnostic) were available for determining if transpiled source code had type-related problems. Now, there is no way to know if the transpiled source is unsound from a type system perspective.

It would be very useful for Deno to provide runtime API access to its type-checking functionality and diagnostics. It's not possible to use `tsc` or tools in its ecosystem for this purpose (primarily because Deno's import specifier syntax is incompatible).

I believe this should be provided directly from the runtime on the `Deno` namespace (instead of as an external module like [`https://github.com/denoland/deno_emit`](https://github.com/denoland/deno_emit)) because the modified version of the compiler that is needed for this task is already so tightly-integrated.

Related:
  - https://github.com/denoland/deno/issues/1739
  - https://github.com/denoland/deno/issues/10731
  - Query for potentially-related open issues: [`check typescript api`](https://github.com/denoland/deno/issues?q=is%3Aissue+is%3Aopen+check+typescript+api)

**Comment by bartlomieju** [comment]
Closing this as out of scope.

A dedicated runtime type-checking API on the `Deno.*` namespace isn't something we plan to add. It's a niche need (no further interest here since 2022), and it runs counter to the current direction of moving type checking onto the stock TypeScript compiler (typescript-go) rather than baking a checker into the runtime namespace. The original blocker mentioned here, that stock `tsc` couldn't understand Deno's import specifiers, is exactly what that migration addresses.

For the use cases this covers today:

- To know whether sources type-check, run `deno check` (it's scriptable via `Deno.Command`, and you can parse its diagnostics output).
- For programmatic type checking, `npm:typescript` works in Deno and gives you the full compiler API and diagnostics.

If a concrete need remains once the stock-TSC work lands, please open a focused issue with the specific API and use case.

**Comment by lishaduck** [comment]
> For programmatic type checking, npm:typescript works in Deno and gives you the full compiler API and diagnostics.

For Node.js projects, yes, but for actual Deno projects? Are you saying that Deno is unforking TS?

For my use-case, Flint is a type-aware Linter that we'd like to natively support Deno projects, so we need compiler diagnostics. Unforking TypeScript would work great for us.

**Comment by bartlomieju** [comment]
> Are you saying that Deno is unforking TS?

Yes
